src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector/CensorshipDetector.cs

```diff
@@ -0,0 +1,187 @@
+                {
+                    _poolAddressCensorshipDetectorHelper.Add(address!, 0);
+                }
+            }
+        }
+
+        _potentiallyCensoredBlocks = new(_cacheSize, _cacheSize, "potentiallyCensoredBlocks");
+        _txPool.NewPending += OnAddingTxToPool;
+        _txPool.RemovedPending += OnRemovingTxFromPool;
+        _txPool.EvictedPending += OnRemovingTxFromPool;
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+    }
+
+    private void OnAddingTxToPool(object? sender, TxPool.TxEventArgs e)
+    {
+        Transaction tx = e.Transaction;
+        if (tx.To is not null && _poolAddressCensorshipDetectorHelper.TryGetValue(tx.To!, out long txSentToAddressCount))
+        {
+            if (txSentToAddressCount == 0)
+            {
+                Metrics.PoolCensorshipDetectionUniqueAddressesCount++;
+            }
+            _poolAddressCensorshipDetectorHelper[tx.To!]++;
+        }
+    }
+
+    private void OnRemovingTxFromPool(object? sender, TxPool.TxEventArgs e)
+    {
+        Transaction tx = e.Transaction;
+        if (tx.To is not null && _poolAddressCensorshipDetectorHelper.TryGetValue(tx.To!, out long txSentToAddressCount) && txSentToAddressCount > 0)
+        {
+            if (txSentToAddressCount == 1)
+            {
+                Metrics.PoolCensorshipDetectionUniqueAddressesCount--;
+            }
+            _poolAddressCensorshipDetectorHelper[tx.To!]--;
+        }
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void Cache(Block block)
+    {
+        Transaction bestTxInPool = _txPool.GetBestTx();
+        long _poolCensorshipDetectionUniqueAddressCount = Metrics.PoolCensorshipDetectionUniqueAddressesCount;
+
+        Transaction bestTxInBlock = block.Transactions[0];
+        SortedDictionary<Address, bool> _blockAddressCensorshipDetectorHelper = [];
+        // Number of unique addresses specified by the user for censorship detection, to which txs are sent in the block
+        long _blockCensorshipDetectionUniqueAddressCount = 0;
+
+        foreach (Transaction tx in block.Transactions)
+        {
+            if (!tx.SupportsBlobs)
+            {
+                if (_poolAddressCensorshipDetectorHelper.TryGetValue(tx.To!, out _) && !_blockAddressCensorshipDetectorHelper.TryGetValue(tx.To!, out _))
+                {
+                    _blockAddressCensorshipDetectorHelper[tx.To!] = true;
+                    _blockCensorshipDetectionUniqueAddressCount++;
+                }
+
+                if (_comparer.Compare(bestTxInBlock, tx) > 0)
+                {
+                    bestTxInBlock = tx;
+                }
+            }
+        }
+
+        bool _highPayingTxCensorship = _comparer.Compare(bestTxInBlock, bestTxInPool) > 0;
+        bool _addressCensorship = _blockCensorshipDetectionUniqueAddressCount * 2 < _poolCensorshipDetectionUniqueAddressCount;
```

Comment:
I kind of dislike this logic:
1. `_poolCensorshipDetectionUniqueAddressCount` doesn't take into account nonces, maybe those transactions aren't ready to be included.
2. `_poolCensorshipDetectionUniqueAddressCount` doesn't take into account basefee or gas premium, maybe other transactions just pay better?

This leads that this can easily identify false positives.